### ISSUE TITLE
Make napari-ome-zarr-navigator work in pixi projects

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/napari-ome-zarr-navigator#29]
Currently, I can't install napari-ome-zarr-navigator in the same pixi project as napari due to conflicts in sub-dependencies of each.

```
pixi init napari-fractal
cd napari-fractal 
pixi add python==3.10
pixi add --pypi napari-ome-zarr-navigator
✔ Added napari-ome-zarr-navigator >=0.1.2, <0.2
pixi add napari
  × failed to solve the pypi requirements of 'default' 'osx-arm64'
  ├─▶ failed to resolve pypi dependencies
  ╰─▶ Because fractal-tasks-core==1.3.2 depends on pydantic>2,<=2.8.2 and pydantic==2.9.2, we can conclude that fractal-tasks-
      core==1.3.2 cannot be used.
      And because napari-ome-zarr-navigator==0.1.2 depends on fractal-tasks-core==1.3.2, we can conclude that napari-ome-zarr-
      navigator==0.1.2 cannot be used.
      And because only napari-ome-zarr-navigator<=0.1.2 is available and you require napari-ome-zarr-navigator>=0.1.2, we can
      conclude that your requirements are unsatisfiable.
```

This is after fixing the numpy dependency issues: https://github.com/fractal-napari-plugins-collection/napari-ome-zarr-navigator/pull/27

napari-ome-zarr-navigator depends on fractal-tasks-core for NGFF library functionality. Fracal-tasks-core depends on pydantic <=2.8.2
When adding napari, something appears to find a conflict within fractal-tasks-core itself. It's not that napari can't coexist with pydantic==2.8.0. But one of the shared dependencies is apparently creating issues here. And some dependency that we don't pin in fractal-tasks-core appears to depend on pydantic==2.9.2 exactly. While this seems like over-aggressive dependency pinning here, it makes the installation a big pain.

In practice, none of what napari-ome-zarr-navigator uses from fractal-tasks-core actually depends on pydantic deep enough where this would matter, because installing it e.g. in conda environments where those dependencies aren't checked doesn't lead to any issues. But we 1) want to work with pixi & 2) want to avoid dependency issues in general.

----

Potential ways to solve this:
1. Switch to using ngio instead of fractal-tasks-core as a dependency in napari-ome-zarr-navigator: https://github.com/fractal-analytics-platform/ngio
2. Relax pydantic constraints in fractal-tasks-core (see discussion here: https://github.com/fractal-analytics-platform/fractal-task-tools/issues/2)
3. Figure out which dependency leads to issue and pin that one

----

Orders of effort:
If there's a simple way to track part 3 down, that would be the easiest. I'm currently failing to find it. Maybe @lorenzocerrone would have an idea?

If not, doing the refactor to switch to ngio is anyway useful, but some work.
Changing fractal-tasks-core dependencies for pydantic would be the hardest for sure

**Comment by jluethi** [comment]
Thanks a lot for digging into this @tcompa ! 

This now works for me to run this napari plugins within pixi. When napari is installed from pypi, we need to manually add pyqt as well. That's one of the conda benefits I think.

Thus, the current way to make this work is:
```
pixi init napari-fractal
cd napari-fractal
pixi add python==3.10
pixi add napari --pypi napari-ome-zarr-navigator
pixi add pyqt
pixi run napari
```

Which leads to the following pixi.toml:
```
[project]
authors = ["jluethi <joel.luethi@uzh.ch>"]
channels = ["conda-forge"]
description = "Add a short description here"
name = "napari-fractal"
platforms = ["osx-arm64"]
version = "0.1.0"

[tasks]

[dependencies]
python = "==3.10"
pyqt = ">=5.15.9,<6"

[pypi-dependencies]
napari = ">=0.5.4, <0.6"
napari-ome-zarr-navigator = ">=0.1.2, <0.2"
```

-------

Very weird that pixi behaves in that way with pydantic dependencies. 
Any idea what are we doing wrong here @lorenzocerrone & @imagejan ?

**Comment by imagejan** [comment]
Just a few comments:

```
$ pixi add napari --pypi napari-ome-zarr-navigator
```
.. is the same as:
```
$ pixi add --pypi napari napari-ome-zarr-navigator
```
.. so both will be installed via PyPI (i.e. the `--pypi` flag affects the whole command).

Pixi solves the conda dependencies first, and then can run into incompatibilities if a PyPI package has different version requirements of a dependency that was already solved in the conda step. (They apply some mapping of pypi->conda packages to solve the transient dependencies via conda if possible.)

---

### Possible explanation

When you `pixi add napari` (via conda), it will have the latest `docstring_parser==0.16` as a dependency (via `magicgui`).
Trying to install `napari-ome-zarr-navigator` will fail because it depends on `docstring-parser=^0.15` (via `fractal-tasks-core`).

---

### Alternative solution

Pinning to `docstring_parser<0.16` in the conda dependencies helps:

```
pixi init
pixi add python==3.10 pydantic==2.8.2 docstring_parser<0.16 napari pyqt
pixi add --pypi napari-ome-zarr-navigator
```

---

**Note**: I noticed that `fractal-tasks-core` on `conda-forge` is at version `1.0.1` (see https://anaconda.org/conda-forge/fractal-tasks-core/files). Any plans to change that? I guess it would also simplify the issue here.

**Comment by imagejan** [comment]
Thanks @tcompa, I [commented](https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/611#issuecomment-2503879258) on the linked issue.

Related question: I can see why fractal tasks can optionally depend on `cellpose` or `scikit-image` to get things done. But why is there a dependency on `stackview`? Aren't tasks supposed to run headlessly?

**Comment by imagejan** [comment]
> We already indirectly dependede on `stackview`, due to `napari-workflows`, but we also included it as a direct dependency so that we could constrain its version, since version 0.9.0 was broken.

Alright, thanks for the insight. So this is actually an issue of `napari-workflows` that depends on `stackview` although it explicitly states that "[it serves as backend](https://github.com/haesleinhuepf/napari-workflows/blob/d57be7573cbf0044c0882a92dc303d34157d22e3/README.md?plain=1#L11)"... 😕

**Comment by jluethi** [comment]
Thanks a lot for the further explanations @imagejan !
Re stackview dependency: The mid-term goal is to move the napari workflows wrapper into its own task package to avoid having to deal with such dependency version issues in tasks-core.

And longer term, library functionality of fractal-tasks-core should either be available in ngio (all OME-Zarr related operations) or in a to-be-extracted package that handles manifest creation etc. => it will become less likely that people depend on tasks-core, but either on ngio or the task building package
